bark_ml/observers/graph_observer.py

```diff
@@ -0,0 +1,554 @@
+# Copyright (c) 2020 fortiss GmbH
+#
+# Authors: Marco Oliva, Silvan Wimmer
+#
+# This work is licensed under the terms of the MIT license.
+# For a copy, see <https://opensource.org/licenses/MIT>.
+
+import numpy as np
+import logging
+import tensorflow as tf
+from gym import spaces
+from typing import Dict
+from collections import OrderedDict
+
+from bark.core.models.dynamic import StateDefinition
+from bark.core.geometry import Distance, Point2d
+from bark.runtime.commons.parameters import ParameterServer
+
+from bark_ml.observers.observer import StateObserver
+
+class GraphObserver(StateObserver):
+  """
+  This observer converts an `ObservsedWorld` instance into
+  a graph-structured observation, consisting of nodes (with
+  features), an adjacency matrix expressing the (directed)
+  connections between nodes in the graph, and edges (with
+  features), expressing the relative information between
+  the agents that constitute these connections.
+  """
+
+  def __init__(self,
+               normalize_observations=True,
+               params=ParameterServer()):
+    """
+    Creates an instance of `GraphObserver`.
+
+    Args:
+      normalize_observations: A boolean value indicating whether
+        the observations returned by the `Observe` function
+        should be normalized into the range of [-1, 1].
+      params: A `ParameterServer` instance that enables further
+        configuration of the observer. Defaults to a new instance.
+    """
+    StateObserver.__init__(self, params)
+    self._normalize_observations = normalize_observations
+    self._logger = logging.getLogger()
+
+    self._agent_limit =\
+      params["ML"]["GraphObserver"]["AgentLimit", 
+      "The maximum number of agents that can be observed.",
+      4]
+
+    self._visibility_radius =\
+      params["ML"]["GraphObserver"]["VisibilityRadius",
+      "The radius in which agent can 'see', i.e. detect other agents.", 
+      50]
+
+    self._add_self_loops =\
+      params["ML"]["GraphObserver"]["SelfLoops", 
+      "Whether each node has an edge pointing to itself.", 
+      False]
+
+    requested_node_attribute_keys =\
+      params["ML"]["GraphObserver"]["EnabledNodeFeatures",
+      "The list of available node features, given by their string key that \
+       the observer should extract from the world and insert into the \
+       observation. For a list of available features, refer to the list \
+       returned by `GraphObserver.available_node_attributes`.",
+      self.available_node_attributes()
+    ]
+
+    self.enabled_node_attribute_keys = self._filter_requested_attributes(
```

Comment:
why do we not use an underscore for this and the following member variables?

Reply:
I could imagine that the properties about which features are enabled might be useful to be read from the outside. But in this case, a read-only property might actually be better. I will add an underscore.

Reply:
as you think! was just a question!